### PR TITLE
Updates to handle multiple search results

### PR DESCRIPTION
**Pull request by jyothishnt**
Tasks done

- If there are multiple results, it should reflect in the json response.
- Also handle when there are no results.
- Create simple html page to show multiple results

https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-2660
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-2661

http://multiple-results.review.ensembl.org/api/id/ENSG00000127720

Screenshot 2024-07-15 at 14 48 10
Screenshot 2024-07-15 at 14 48 22

### CODE REVIEW

File: app/api/resources/resolver_view.py [hunk 1]
```diff
@@ -14,36 +14,80 @@
...
+    return HTMLResponse(generate_html_content(resolved_genome_info))
 
-    return RedirectResponse(url)
+def generate_html_content(resolved_genome_info):
+    # Create a simple HTML page with a list of URLs
+  html_content = "<html><body><h1>Resolved URLs</h1><table border='1' style='border-collapse: collapse;'><tr><th>Assembly Name</th><th>Accession ID</th><th>Ensembl URL</th></tr>"
+  for genome_info in resolved_genome_info:
+      html_content += f'<tr><td>{genome_info["assembly_name"]}</td><td>{genome_info["accession_id"]}</td><td><a href="{genome_info["resolved_url"]}">{genome_info["resolved_url"]}</a></td></tr>'
+  html_content += "</table></body></html>"
+  return html_content
```
Comment (azangru) [inline app/api/resources/resolver_view.py:93]: I would probably add a templating engine (e.g. jinja) here, and move the css into its own static file for maintainability.

File: app/api/resources/resolver_view.py [hunk 2]
```diff
@@ -14,36 +14,80 @@
...
+  meta_results = {}
+  resolved_genome_info = []
 
-    if (app == "entity-viewer"):
-      url=f"{ENSEMBL_URL}/{app}/{genome_id}/{type}:{stable_id}"
+  for match in matches:
+    genome_id = match.get("genome_id")
+    try:
+      with meta_session.get(url=f"{ENSEMBL_URL}/api/metadata/genome/{genome_id}/details") as meta_response:
+        meta_response.raise_for_status()
```
Comment (azangru) [inline app/api/resources/resolver_view.py:52]: We aren't worried by the number of requests to the metadata api?

File: app/api/resources/resolver_view.py [hunk 3]
```diff
@@ -14,36 +14,80 @@
...
-    if not matches:
-      return {"response": "No matches found"}
 
-    genome_id = matches[0].get("genome_id")
+  if not matches:
+    return response_error_handler({"status": 404})
+
+  # Get assembly information
+  meta_session = requests.Session()
+  meta_results = {}
```
Comment (azangru) [inline app/api/resources/resolver_view.py:45]: Why do you need it here?

File: app/api/resources/resolver_view.py [hunk 4]
```diff
@@ -14,36 +14,80 @@
...
+  resolved_genome_info = []
```
Comment (azangru) [inline app/api/resources/resolver_view.py:46]: `resolved_pages` perhaps? This is the accumulator of what the resolver resolves. It does not resolve genomes.

File: app/api/resources/resolver_view.py [hunk 5]
```diff
@@ -14,36 +14,80 @@
...
+        "accession_id": meta_results["assembly"]["accession_id"],
+        "assembly_name": meta_results["assembly"]["name"],
+        "species_scientific_name": meta_results["scientific_name"],
+        "taxonomy_id": meta_results["taxonomy_id"],
+        "resolved_url": url
+      }
+
+      resolved_genome_info.append(meta)
+    
+  if request.headers.get("content-type") == "application/json":
```
Comment (kamaldodiya) [inline app/api/resources/resolver_view.py:79]: The header to check here would be Accept not content type.

File: app/api/resources/resolver_view.py [hunk 6]
```diff
@@ -14,36 +14,80 @@
...
+    return resolved_genome_info
```
Comment (kamaldodiya) [inline app/api/resources/resolver_view.py:80]: `resolved_genome_info` is list here and not JSON if that is what is expected here. May be FastAPI could be converting it to JSON but did you check if it is the case ?

**Review by kamaldodiya — CHANGES_REQUESTED** [review]

File: app/api/resources/resolver_view.py (hunk 2 above)
Comment (jyothishnt) [reply app/api/resources/resolver_view.py:52]: Hmm good point. But other than this how would we get the assembly info? Or else we would need an endpoint that takes multiple genome ids and send back metadata.

File: app/api/resources/resolver_view.py (hunk 6 above)
Comment (jyothishnt) [reply app/api/resources/resolver_view.py:80]: Fastapi returns a Json.

File: app/api/resources/resolver_view.py (hunk 2 above)
Comment (jyothishnt) [reply app/api/resources/resolver_view.py:52]: We decided in code review to go with web metadata api for now.

**Review by kamaldodiya — APPROVED** [review]

File: app/api/resources/templates/interstitial.j2
Comment (azangru) [inline app/api/resources/templates/interstitial.j2:1]: "interstitial" is probably the wrong name for this file (as well as for the css file). An interstitial means something in between; like the screens we have on the beta site that separate you from getting into the app. But here, in the context of this application, there is nothing around this page; nothing for it to be an interstitial for. It is probably "links", or "resolved_links", or "search_results", or something of that nature.

File: app/api/resources/templates/interstitial.j2 [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<!doctype html>
```
Comment (azangru) [inline app/api/resources/templates/interstitial.j2:1]: - Where is the `html` tag? Also, let's have a language on it.
- Let's at least have a page title.

File: app/core/config.py [hunk 8]
```diff
@@ -25,7 +25,7 @@
 from .logging import InterceptHandler
 
 VERSION = "0.0.0"
-API_PREFIX = "/"
+API_PREFIX = "/api"
```
Comment (azangru) [inline app/core/config.py:28]: What is the full url for the resolver going to be?

File: app/api/resources/resolver_view.py [hunk 9]
```diff
@@ -1,49 +1,96 @@
...
 from fastapi import APIRouter, Request
 from typing import Optional
-from fastapi.responses import RedirectResponse
+from fastapi.responses import RedirectResponse, HTMLResponse
 from loguru import logger
 import requests, logging
+import aiohttp
+from dotenv import load_dotenv
+from jinja2 import Environment, FileSystemLoader
+from jinja2 import Template
```
Comment (azangru) [inline app/api/resources/resolver_view.py:10]: Why two lines (9, 10)?

File: app/api/resources/resolver_view.py [hunk 10]
```diff
@@ -1,49 +1,96 @@
...
+  else:
+    return HTMLResponse(generate_html_content(results))
 
-    return RedirectResponse(url)
+def generate_html_content(results):
+    # Create a simple HTML page with a list of URLs
+  load_dotenv()
+  CURR_DIR = os.path.dirname(os.path.abspath(__file__))
+  env = Environment(loader=FileSystemLoader(os.path.join(CURR_DIR,"templates")))
+  interstitial_template = env.get_template("interstitial.j2")
```
Comment (azangru) [inline app/api/resources/resolver_view.py:94]: In examples in FastAPI docs, they use the html extension for jinja templates. Do we prefer a j2 extension?

File: app/core/config.py (hunk 8 above)
Comment (jyothishnt) [reply app/core/config.py:28]: Make API_PREFIX empty

File: app/api/resources/templates/interstitial.j2
Comment (jyothishnt) [reply app/api/resources/templates/interstitial.j2:1]: Changing to `search_results`

File: app/api/resources/resolver_view.py (hunk 9 above)
Comment (jyothishnt) [reply app/api/resources/resolver_view.py:10]: Removing, as I am not using it anywhere.

File: app/api/resources/resolver_view.py (hunk 10 above)
Comment (jyothishnt) [reply app/api/resources/resolver_view.py:94]: Although we used .j2 in [ticket manager](https://gitlab.ebi.ac.uk/ensembl-web/ensembl_ticket_manager/-/tree/master/app/api/resources/mail_manager/templates?ref_type=heads)
I prefer using .html. If thats not a problem then I will update this. @Mehrnaz-Charkhchi @kamaldodiya ?

Comment (Mehrnaz-Charkhchi) [reply app/api/resources/resolver_view.py:94]: @jyothishnt It doesn't seem to be making a difference according to the docs so I would say go with whatever you're comfortable with.

File: app/main.py [hunk 11]
```diff
@@ -39,3 +39,4 @@ def get_application() -> FastAPI:
 
 
 app = get_application()
+app.mount("/static/css", StaticFiles(directory="static/css"), name="css")
```
Comment (azangru) [inline app/main.py:42]: Shouldn't we just mount the `/static` directory as a whole? Why narrow it down to `/static/css`?

File: app/api/resources/resolver_view.py [hunk 12]
```diff
@@ -1,49 +1,95 @@
...
+        "accession_id": meta_results["assembly"]["accession_id"],
+        "assembly_name": meta_results["assembly"]["name"],
+        "species": meta_results["scientific_name"] if meta_results["scientific_name"] else meta_results["common_name"],
+        "type": meta_results["type"],
+        "resolved_url": url
+      }
+
+      results.append(meta)
+
+  # return request.headers
```
Comment (azangru) [inline app/api/resources/resolver_view.py:78]: ✂️

**Review by azangru — APPROVED** [review]

File: app/main.py [hunk 13]
```diff
@@ -39,3 +39,4 @@ def get_application() -> FastAPI:
 
 
 app = get_application()
+app.mount("/static", StaticFiles(directory="static"), name="static files")
```
Comment (jyothishnt) [inline app/main.py:42]:
```suggestion
app.mount("/static", StaticFiles(directory="static"), name="static_files")

```

**Pull request merged**